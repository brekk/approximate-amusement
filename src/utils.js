import {
  propEq,
  ifElse,
  slice,
  last,
  reject,
  memoizeWith,
  identity as I,
  curry,
  cond,
  sortBy,
  pipe,
  prop,
  T as otherwise
} from 'ramda'
import { SENTIMENTS } from './constants'

const memo = memoizeWith(I)

export const within = curry(
  (lower, upper, x) => x <= upper && x > lower
)
// scale number values from `randoNum` down to array indices
// TODO, this `cond` is unwieldy, change it if we have time
const sentimeter = x => () => SENTIMENTS[x]
export const enthusiasm = memo(
  cond([
    [within(-110, -90), sentimeter(0)],
    [within(-90, -80), sentimeter(1)],
    [within(-80, -70), sentimeter(2)],
    [within(-70, -60), sentimeter(3)],
    [within(-60, -50), sentimeter(4)],
    [within(-50, -40), sentimeter(5)],
    [within(-40, -30), sentimeter(6)],
    [within(-30, -20), sentimeter(7)],
    [within(-20, -10), sentimeter(8)],
    [within(-10, 0), sentimeter(9)],
    [within(0, 10), sentimeter(10)],
    [within(10, 20), sentimeter(11)],
    [within(20, 30), sentimeter(12)],
    [within(30, 40), sentimeter(13)],
    [within(40, 50), sentimeter(14)],
    [within(50, 60), sentimeter(15)],
    [within(60, 70), sentimeter(16)],
    [within(70, 80), sentimeter(17)],
    [within(80, 90), sentimeter(18)],
    [within(90, 110), sentimeter(19)],
    [otherwise, sentimeter(10)]
  ])
)

// generate a random number
// NB: Math.random is very non-random, but fine for these porpoises 🐬
// randoNum :: x -> Number
export const randoNum = () => {
  const x = Math.floor(Math.random() * 100)
  if (Math.round(Math.random() * 1) === 1) {
    return x * -1
  }
  return x
}

export const tackOn = curry((fn, z) => [z, fn(z)])
export const sortByScore = sortBy(pipe(prop('score'), z => z * -1))

export const isTopValue = curry((knownValues, x) =>
  ifElse(
    z => z.length < 10,
    () => true,
    pipe(
      sortByScore,
      slice(0, 10),
      last,
      prop('score'),
      lowestTopScore => x > lowestTopScore
    )
  )(knownValues)
)

export const realScoresOnly = reject(propEq('fake', true))
