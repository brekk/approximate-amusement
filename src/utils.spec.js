import { pipe, map, range } from 'ramda'
// import { trace } from 'xtrace'
import { tackOn, enthusiasm, randoNum, within } from './utils'

test('within', () => {
  expect(within(0, 10, -1)).toBeFalsy()
  expect(within(0, 10, 0)).toBeFalsy()
  expect(within(0, 10, 5)).toBeTruthy()
  expect(within(0, 10, 10)).toBeTruthy()
  expect(within(0, 10, 11)).toBeFalsy()
})

test('randoNum', () => {
  pipe(range(0), map(randoNum), ([a, b, c]) =>
    expect(a !== b && b !== c && a !== c).toBeTruthy()
  )(3)
})

test('enthusiasm', () => {
  const output = pipe(
    range(-10),
    map(z => z * 10),
    map(tackOn(enthusiasm))
  )(11)
  expect(output).toEqual([
    [-100, 'Abysmal!'],
    [-90, 'Abysmal!'],
    [-80, 'Terrible!'],
    [-70, 'Bad!'],
    [-60, 'Poor!'],
    [-50, 'Weak.'],
    [-40, 'Pedestrian.'],
    [-30, 'Pretty poor.'],
    [-20, 'You can do better.'],
    [-10, 'Not good.'],
    [0, 'ðŸ˜•'],
    [10, 'Meh.'],
    [20, 'ðŸ™‚'],
    [30, 'Alright.'],
    [40, 'Okay.'],
    [50, 'Decent.'],
    [60, 'Good.'],
    [70, 'Nice.'],
    [80, 'Great!'],
    [90, 'Awesome!'],
    [100, 'Fantastic!']
  ])
})
