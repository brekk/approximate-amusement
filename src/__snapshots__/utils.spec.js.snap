// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enthusiasm 1`] = `
Array [
  Array [
    -100,
    "Abysmal!",
  ],
  Array [
    -90,
    "Abysmal!",
  ],
  Array [
    -80,
    "Terrible!",
  ],
  Array [
    -70,
    "Bad!",
  ],
  Array [
    -60,
    "Poor!",
  ],
  Array [
    -50,
    "Weak.",
  ],
  Array [
    -40,
    "Pedestrian.",
  ],
  Array [
    -30,
    "Pretty poor.",
  ],
  Array [
    -20,
    "You can do better.",
  ],
  Array [
    -10,
    "Not good.",
  ],
  Array [
    0,
    "Do better.",
  ],
  Array [
    10,
    "Meh.",
  ],
  Array [
    20,
    "Alright.",
  ],
  Array [
    30,
    "Keep trying.",
  ],
  Array [
    40,
    "Okay.",
  ],
  Array [
    50,
    "Decent.",
  ],
  Array [
    60,
    "Good.",
  ],
  Array [
    70,
    "Nice.",
  ],
  Array [
    80,
    "Great!",
  ],
  Array [
    90,
    "Awesome!",
  ],
  Array [
    100,
    "Fantastic!",
  ],
]
`;
